examples/token/src/lib.rs

```diff
@@ -0,0 +1,324 @@
+            };
+            transfer(state, params);
+            None
+        }
+        5 => {
+            let addr: Address = match params_cbor(params_id) {
+                Ok(addr) => addr,
+                Err(err) => abort!(ErrIllegalArgument, "failed to parse address: {:?}", err),
+            };
+            Some(balance_of(state, addr))
+        }
+
+        // TODO this exit code should move to actor space given that dispatch
+        //  is now conducted in-actor.
+        _ => abort!(SysErrInvalidMethod, "unrecognized method"),
+    };
+
+    // Insert the return data block if necessary, and return the correct
+    // block ID.
+    match ret {
+        None => NO_DATA_BLOCK_ID,
+        Some(v) => match sdk::ipld::put_block(DAG_CBOR, v.bytes()) {
+            Ok(id) => id,
+            Err(err) => abort!(ErrSerialization, "failed to store return value: {}", err),
+        },
+    }
+}
+
+/// Returns the token name. Method number 1.
+#[inline]
+pub fn name(state: State) -> RawBytes {
+    encode!(state.name)
+}
+
+/// Returns the symbol. Method number 2.
+#[inline]
+pub fn symbol(state: State) -> RawBytes {
+    encode!(state.symbol)
+}
+
+/// Returns the maximum supply. Method number 3.
+#[inline]
+pub fn max_supply(state: State) -> RawBytes {
+    encode!(BigIntSer(&state.max_supply))
+}
+
+/// The input parameters for a transfer.
+#[derive(Serialize_tuple, Deserialize_tuple, Clone, Debug)]
+pub struct TransferParams {
+    pub recipient: Address,
+    #[serde(with = "bigint_ser")]
+    pub amount: TokenAmount,
+}
+
+impl Cbor for TransferParams {}
+
+/// Transfer a token amount.
+pub fn transfer(mut state: State, params: TransferParams) {
+    // Load the balances HAMT.
+    // TODO Using BitIntDe because it's both Ser and De; this is a misnomer and
+    //  we should fix it.
+    let mut balances =
+        match Hamt::<Blockstore, BigIntDe, ActorID>::load(&state.balances, Blockstore) {
+            Ok(map) => map,
+            Err(err) => abort!(ErrIllegalState, "failed to load balances hamt: {:?}", err),
+        };
+
+    // Load the sender's balance.
+    let sender_id = fvm_sdk::message::caller();
+    let mut sender_bal = match balances.get(&sender_id) {
+        Ok(Some(bal)) => bal.clone(),
+        Ok(None) => BigIntDe(TokenAmount::zero()),
+        Err(err) => abort!(ErrIllegalState, "failed to get balance: {:?}", err),
+    };
+
+    // Sender has insufficient balance.
+    if sender_bal.0 < params.amount {
+        abort!(ErrInsufficientFunds, "sender has insufficient balance")
+    }
+
+    // Resolve the recipient into an ID address.
+    // TODO See addressing section on module docs.
+    let recipient_id = match fvm_sdk::actor::resolve_address(&params.recipient) {
+        Some(id) => id,
+        None => abort!(ErrIllegalArgument, "failed to resolve address"),
+    };
+
+    // Forbid sends to self.
+    if sender_id == recipient_id {
```

Comment:
Yes, I don't think denying the transfer is an appropriate fix. It's interesting that ERC-20 doesn't specifically address this case, but reading between the lines of "must emit a Transfer event even for values of zero" I think the intention would be to support this. The common implementations ([1](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/9b3710465583284b8c4c5d2245749246bb2e0094/contracts/token/ERC20/ERC20.sol#L60), [2](https://github.com/ConsenSys/Tokens/blob/fdf687c69d998266a95f15216b1955a4965a0a6d/contracts/eip20/EIP20.sol#L39)) both do, by doing a safe sequence of operations of read-modify-write each balance in turn, rather than interleaved.